File: e-java/src/main/java/e/codec/Encoder.java
package e.codec;

/**
 * Typeclass defining how to encode an input value to an output value
 *
 * @param <I> Type of input
 * @param <O> Type of output
 */
public interface Encoder<I, O> {
    /**
     * Encodes an input
     *
     * @param input An input
     *
     * @return Encoded output
     */
    O encode(I input);
}

File: e-kotlin/src/main/kotlin/e/kotlin/codec/Encoder.kt
package e.kotlin.codec

/**
 * Typeclass defining how to encode an input value to an output value
 *
 * @param I Type of input
 * @param O Type of output
 */
interface Encoder<in I, out O> {
  /**
   * Encodes an input
   *
   * @param input An input
   *
   * @return Encoded output
   */
  fun encode(input: I): O
}
